bundles/org.openhab.binding.upb/src/main/java/org/openhab/binding/upb/handler/SerialPIMHandler.java

```diff
@@ -0,0 +1,180 @@
+    private volatile ScheduledFuture<?> futSerialPortInit;
+
+    public SerialPIMHandler(final Bridge thing, final SerialPortManager serialPortManager) {
+        super(thing);
+        this.serialPortManager = serialPortManager;
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Initializing Serial UPB PIM {}.", getThing().getUID());
+        super.initialize();
+
+        final String portId = (String) getConfig().get(Constants.CONFIGURATION_PORT);
+        if (portId == null || portId.isEmpty()) {
+            logger.error("serial port is not set");
+            return;
+        }
+
+        futSerialPortInit = scheduler.schedule(() -> openSerialPort(portId), SERIAL_PORT_OPEN_INIT_DELAY_MS,
+                TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void dispose() {
+        if (futSerialPortInit != null) {
+            futSerialPortInit.cancel(true);
+        }
+        futSerialPortInit = null;
+        if (receiveThread != null) {
+            receiveThread.terminate();
+            try {
+                receiveThread.join(1000);
+            } catch (final InterruptedException e) {
+                // ignore
+            }
+            receiveThread = null;
+        }
+        logger.info("Stopped UPB serial handler");
+        super.dispose();
+    }
+
+    private void openSerialPort(final String portId) {
+        try {
+            final SerialPort serialPort = tryOpenSerialPort(portId);
+            if (serialPort == null) {
+                futSerialPortInit = scheduler.schedule(() -> openSerialPort(portId), SERIAL_PORT_OPEN_RETRY_DELAY_MS,
+                        TimeUnit.MILLISECONDS);
+                return;
+            }
+            logger.debug("Starting receive thread");
+            receiveThread = new SerialIoThread(serialPort, this);
+            receiveThread.setName("upb-serial-reader");
+            // Once the receiver starts, it may set the PIM status to ONLINE
+            // so we must ensure all initialization is finished at that point.
+            receiveThread.start();
+            updateStatus(ThingStatus.ONLINE);
+        } catch (final Exception e) {
+            logger.error("failed to open serial port", e);
+        }
+    }
+
+    private SerialPort tryOpenSerialPort(final String portId) {
+        logger.info("opening serial port {}", portId);
+        final SerialPortIdentifier portIdentifier = serialPortManager.getIdentifier(portId);
+        if (portIdentifier == null) {
+            logger.error("port not found: {}", portId);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                    Constants.OFFLINE_SERIAL_EXISTS);
+            return null;
+        }
+
+        final SerialPort serialPort;
+        try {
+            serialPort = portIdentifier.open("org.openhab.binding.upb", 1000);
+        } catch (final PortInUseException e) {
+            logger.error("cannot open serial port", e);
```

Comment:
remove this logging statement, the user is already notified through a change in the Thing status.